### PR TITLE
dial node/address with bad server cert

### PR DESCRIPTION
**Pull request by bryanchriswhite**
This PR fixes the indefinite hang in the case where the remote peer (server) has an invalid tls configuration. It also adds tests to `transport` for the cases where client/server tls configurations are in/valid.

### CODE REVIEW

**Comment by bryanchriswhite** [comment]
I'm not sure how best to deal with the `FetchPeerIdentity` problem. Because of the use of a custom dialer, the grpc peer doesn't have `authInfo`. We receive it in the dialer function when calling `tlsConfig.ClientHandshake` but I'm not sure how best to persist it from there to the places where we currently need it; i.e. `FetchPeerIdentity`.

**Comment by egonelbre** [comment]
I know this can be done via using custom credentials (see https://github.com/storj/storj/pull/492/files). Not sure whether there's an easier way.

File: pkg/transport/transport.go [hunk 1]
```diff
@@ -64,9 +64,11 @@ func (transport *Transport) DialNode(ctx context.Context, node *pb.Node, opts ..
 		return nil, Error.New("no address")
 	}
 
-	// add ID of node we are wanting to connect to
-	dialOpt := transport.tlsOpts.DialOption(node.Id)
-	options := append([]grpc.DialOption{dialOpt, grpc.WithBlock(), grpc.FailOnNonTempDialError(true)}, opts...)
+	options := append([]grpc.DialOption{
+		transport.tlsOpts.DialOption(ctx, node.Id),
+		grpc.WithBlock(),
+		grpc.FailOnNonTempDialError(true),
```
Comment (jtolds) [inline pkg/transport/transport.go]: should this still be here?

File: pkg/transport/transport.go [hunk 2]
```diff
@@ -89,8 +91,11 @@ func (transport *Transport) DialNode(ctx context.Context, node *pb.Node, opts ..
 func (transport *Transport) DialAddress(ctx context.Context, address string, opts ...grpc.DialOption) (conn *grpc.ClientConn, err error) {
 	defer mon.Task()(&ctx)(&err)
 
-	dialOpt := transport.tlsOpts.DialOption(storj.NodeID{})
-	options := append([]grpc.DialOption{dialOpt, grpc.WithBlock(), grpc.FailOnNonTempDialError(true)}, opts...)
+	options := append([]grpc.DialOption{
+		transport.tlsOpts.DialOption(ctx, storj.NodeID{}),
+		grpc.WithBlock(),
+		grpc.FailOnNonTempDialError(true),
```
Comment (jtolds) [inline pkg/transport/transport.go]: same

File: pkg/transport/transport.go (hunk 1 above)
Comment (bryanchriswhite) [reply pkg/transport/transport.go]: Yes

File: pkg/transport/transport.go (hunk 2 above)
Comment (bryanchriswhite) [reply pkg/transport/transport.go]: We still need these for the same reasons as before. The change to transport credentials just ensures that we get a non-temporary error that will cause grpc to stop trying to reconnect.

File: pkg/peertls/tlsopts/tls.go [hunk 3]
```diff
@@ -37,25 +45,46 @@ func (opts *Options) ServerOption() grpc.ServerOption {
 }
 
 // DialOption returns a grpc `DialOption` for making outgoing connections
-// to the node with this peer identity
-// id is an optional id of the node we are dialing
-func (opts *Options) DialOption(id storj.NodeID) grpc.DialOption {
+// to the node with this peer identity.
+// id is an optional id of the node we are dialing.
+func (opts *Options) DialOption(ctx context.Context, id storj.NodeID) grpc.DialOption {
```
Comment (bryanchriswhite) [inline pkg/peertls/tlsopts/tls.go]: Don't need to receive `ctx`

File: pkg/peertls/tlsopts/tls.go [hunk 4]
```diff
@@ -37,25 +45,46 @@ func (opts *Options) ServerOption() grpc.ServerOption {
...
 		),
 	}
+}
 
-	return grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig))
+// ClientHandshake returns a non-temporary error if an error is returned from
+// the underlying `ClientHandshake` call.
+func (tc *TransportCredentials) ClientHandshake(ctx context.Context, authority string, conn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	tlsConn, authInfo, err := tc.TransportCredentials.ClientHandshake(ctx, authority, conn)
+	if err != nil {
```
Comment (bryanchriswhite) [inline pkg/peertls/tlsopts/tls.go]: Maybe we should make sure this error is what we think it is to allow for proper functionality on the event of a temporary error (whatever that might be).

File: examples/grpc-debug/main.go [hunk 5]
```diff
@@ -40,7 +40,7 @@ func main() {
 
 	dialOption := clientOptions.DialOption(storj.NodeID{})
 
-	conn, err := grpc.Dial(*targetAddr, dialOption)
+	conn, err := grpc.Dial(*targetAddr, dialOption, grpc.WithInsecure())
```
Comment (egonelbre) [inline examples/grpc-debug/main.go]: Add a note why it's fine to have `grpc.WithInsecure` here.

File: pkg/peertls/tlsopts/tls.go [hunk 6]
```diff
@@ -15,8 +18,14 @@ import (
 	"storj.io/storj/pkg/storj"
 )
 
+// TransportCredentials wraps an embedded credentials.TransportCredentials.
+// It decorates the `ClientHandshake` method.
+type TransportCredentials struct {
```
Comment (egonelbre) [inline pkg/peertls/tlsopts/tls.go]: Should this be in `pkg/transport` instead?

File: pkg/peertls/utils.go [hunk 7]
```diff
@@ -22,6 +22,32 @@ import (
 	"storj.io/storj/pkg/pkcrypto"
 )
 
+// NonTemporaryError is an error with a `Temporary` method which always returns false.
+// It is intended for use with grpc.
+//
+// (see https://godoc.org/google.golang.org/grpc#WithDialer
+// and https://godoc.org/google.golang.org/grpc#FailOnNonTempDialError).
+type NonTemporaryError struct {
```
Comment (egonelbre) [inline pkg/peertls/utils.go]: Same here, should this be in `pkg/transport`?

File: pkg/transport/transport_test.go [hunk 8]
```diff
@@ -64,34 +86,140 @@ func TestDialNode(t *testing.T) {
...
+	defer ctx.Check(planet.Shutdown)
+
+	whitelistPath, err := planet.WriteWhitelist()
+	require.NoError(t, err)
+
+	planet.Start(ctx)
+
+	client := planet.StorageNodes[0].Transport
+	signedIdentities := testplanet.NewPregeneratedSignedIdentities()
+	ident, err := signedIdentities.NewIdentity()
```
Comment (egonelbre) [inline pkg/transport/transport_test.go]: `testplanet.PregeneratedSignedIdentity(0)`

File: pkg/transport/transport_test.go [hunk 9]
```diff
@@ -27,11 +30,30 @@ func TestDialNode(t *testing.T) {
...
+	whitelistPath, err := planet.WriteWhitelist()
+	require.NoError(t, err)
+
 	planet.Start(ctx)
 
 	client := planet.StorageNodes[0].Transport
 
-	{ // DialNode with invalid targets
+	unsignedIdentities := testplanet.NewPregeneratedIdentities()
+	unsignedIdent, err := unsignedIdentities.NewIdentity()
```
Comment (egonelbre) [inline pkg/transport/transport_test.go]: `testplanet.PregeneratedIdentity(0)`

File: pkg/peertls/tlsopts/tls.go [hunk 10]
```diff
@@ -15,8 +18,14 @@ import (
 	"storj.io/storj/pkg/storj"
 )
 
+// TransportCredentials wraps an embedded credentials.TransportCredentials.
+// It decorates the `ClientHandshake` method.
```
Comment (egonelbre) [inline pkg/peertls/tlsopts/tls.go]: This comment could be better. Maybe `TransportCredentials wraps credential.TransportCredentials ensuring invalid certificates don't hang dialing.`

The information about embedding and decorating usually don't offer insight why something exists.

File: examples/grpc-debug/main.go (hunk 5 above)
Comment (bryanchriswhite) [reply examples/grpc-debug/main.go]: This should be removed... good catch!

File: pkg/peertls/tlsopts/tls.go (hunk 10 above)
Comment (bryanchriswhite) [reply pkg/peertls/tlsopts/tls.go]: :+1:

File: pkg/transport/transport_test.go (hunk 9 above)
Comment (bryanchriswhite) [reply pkg/transport/transport_test.go]: :+1:

File: pkg/transport/transport_test.go (hunk 8 above)
Comment (bryanchriswhite) [reply pkg/transport/transport_test.go]: :+1:

Comment (bryanchriswhite) [reply pkg/transport/transport_test.go]: I will eventually stop doing this :p

File: pkg/peertls/utils.go (hunk 7 above)
Comment (bryanchriswhite) [reply pkg/peertls/utils.go]: :+1:

File: pkg/peertls/tlsopts/tls.go (hunk 6 above)
Comment (bryanchriswhite) [reply pkg/peertls/tlsopts/tls.go]: :+1:

File: pkg/peertls/utils.go (hunk 7 above)
Comment (egonelbre) [inline pkg/peertls/utils.go]: Can we implement this as 
```
type NonTemporaryError struct { error }
func (err NonTemporaryError) Temporary() bool { return false }
```

Comment (bryanchriswhite) [reply pkg/peertls/utils.go]: yet another good idea

File: pkg/transport/transport.go (hunk 2 above)
Comment (bryanchriswhite) [reply pkg/transport/transport.go]: The non-temporary error change has now been pushed into the peer certificate verification function wrapper.

File: pkg/transport/transport.go (hunk 1 above)
Comment (jtolds) [reply pkg/transport/transport.go]: oh, sorry, for some reason i thought it was FailOnTempDialError

**Pull request merged**